File: Sources/ManageArmsCore/Frontmatter.swift
import Foundation

/// SKILL.md の frontmatter。一覧に必要なのは name と description だけ（DESIGN.md 9 章）。
public struct Frontmatter: Equatable, Sendable {
    public var name: String?
    public var description: String?
    /// Subagent の判定に使う。SKILL.md には無く、agent の .md にはある（DESIGN.md 6 章）。
    public var tools: String?
}

public enum FrontmatterResult: Equatable, Sendable {
    case parsed(Frontmatter)
    /// 先頭が `---` で始まっていない。
    case missing
    /// 読み取った範囲内に閉じ `---` が無い。4 KB しか読まない設計の副作用なので
    /// 「description 無し」と混同せず、UI では明示する（10.1）。
    case truncated
}

public enum FrontmatterParser {
    /// 先頭からこのバイト数だけ読む。DESIGN.md 9 章「frontmatter だけ読む」。
    public static let headBytes = 4096

    /// ファイルの先頭 4 KB だけを読む。本文はメモリに載せない。
    public static func read(_ url: URL) -> FrontmatterResult {
        guard let handle = try? FileHandle(forReadingFrom: url) else { return .missing }
        defer { try? handle.close() }
        let data = (try? handle.read(upToCount: headBytes)) ?? Data()
        return parse(String(decoding: data, as: UTF8.self))
    }

    public static func parse(_ text: String) -> FrontmatterResult {
        // CRLF と BOM を先に潰す
        let normalized = text
            .replacingOccurrences(of: "\r\n", with: "\n")
            .trimmingCharacters(in: CharacterSet(charactersIn: "\u{FEFF}"))
        var lines = normalized.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)

        guard lines.first?.trimmingCharacters(in: .whitespaces) == "---" else { return .missing }
        lines.removeFirst()

        // 閉じ `---` を探す。本文中の `---` に引っかからないよう、行全体が `---` のものだけ。
        guard let end = lines.firstIndex(where: { $0.trimmingCharacters(in: .whitespaces) == "---" })
        else { return .truncated }

        return .parsed(parseBody(Array(lines[..<end])))
    }

    static func parseBody(_ lines: [String]) -> Frontmatter {
        var result = Frontmatter()
        var index = 0
        while index < lines.count {
            let line = lines[index]
            index += 1
            guard !line.hasPrefix(" "), !line.hasPrefix("\t"),
                  let colon = line.firstIndex(of: ":") else { continue }

            let key = String(line[..<colon]).trimmingCharacters(in: .whitespaces)
            var value = String(line[line.index(after: colon)...]).trimmingCharacters(in: .whitespaces)

            // YAML ブロックスカラー（`>` `>-` `|` `|-`）。実在するスキルの大半がこの形式。
            if value.first == ">" || value.first == "|" {
                let fold = value.first == ">"
                var block: [String] = []
                while index < lines.count {
                    let next = lines[index]
                    if !next.isEmpty, !next.hasPrefix(" "), !next.hasPrefix("\t") { break }
                    block.append(next.trimmingCharacters(in: .whitespaces))
                    index += 1
                }
                while block.last?.isEmpty == true { block.removeLast() }
                value = block.joined(separator: fold ? " " : "\n")
            } else {
                value = unquote(value)
            }

            switch key {
            case "name":        result.name = value.isEmpty ? nil : value
            case "description": result.description = value.isEmpty ? nil : value
            case "tools":       result.tools = value.isEmpty ? nil : value
            default:            break
            }
        }
        return result
    }

    static func unquote(_ s: String) -> String {
        guard s.count >= 2, let first = s.first, let last = s.last,
              first == last, first == "\"" || first == "'" else { return s }
        return String(s.dropFirst().dropLast())
    }
}

File: core/frontmatter.ts
/** SKILL.md の frontmatter。一覧に必要なのは name と description だけ。 */
export type Frontmatter = {
  name?: string;
  description?: string;
  /** Subagent の判定に使う。SKILL.md には無く、Subagent の .md にはある。 */
  tools?: string;
};

export type FrontmatterResult =
  | { status: "parsed"; matter: Frontmatter }
  /** 先頭が `---` で始まっていない。 */
  | { status: "missing" }
  /**
   * 読み取った範囲内に閉じ `---` が無い。4 KB しか読まない設計の副作用なので
   * 「description 無し」と混同せず、UI では区別して出す。
   */
  | { status: "truncated" };

/** 先頭からこのバイト数だけ読む。本文はメモリに載せない。 */
export const HEAD_BYTES = 4096;

export function parse(text: string): FrontmatterResult {
  const lines = text.replace(/\r\n/g, "\n").replace(/^﻿/, "").split("\n");
  if (lines[0]?.trim() !== "---") return { status: "missing" };
  const body = lines.slice(1);
  // 本文中の `---` に引っかからないよう、行全体が `---` のものだけを閉じとみなす。
  const end = body.findIndex(line => line.trim() === "---");
  if (end < 0) return { status: "truncated" };
  return { status: "parsed", matter: parseBody(body.slice(0, end)) };
}

function parseBody(lines: string[]): Frontmatter {
  const matter: Frontmatter = {};
  for (let index = 0; index < lines.length; index++) {
    const line = lines[index];
    if (line.startsWith(" ") || line.startsWith("\t")) continue;
    const colon = line.indexOf(":");
    if (colon < 0) continue;

    const key = line.slice(0, colon).trim();
    let value = line.slice(colon + 1).trim();

    // YAML ブロックスカラー（`>` `>-` `|` `|-`）。実在するスキルの大半がこの形式。
    if (value.startsWith(">") || value.startsWith("|")) {
      const fold = value.startsWith(">");
      const block: string[] = [];
      while (index + 1 < lines.length) {
        const next = lines[index + 1];
        if (next !== "" && !next.startsWith(" ") && !next.startsWith("\t")) break;
        block.push(next.trim());
        index++;
      }
      while (block.length > 0 && block[block.length - 1] === "") block.pop();
      value = block.join(fold ? " " : "\n");
    } else {
      value = unquote(value);
    }

    if (value === "") continue;
    if (key === "name") matter.name = value;
    else if (key === "description") matter.description = value;
    else if (key === "tools") matter.tools = value;
  }
  return matter;
}

function unquote(value: string): string {
  const quote = value[0];
  if (value.length >= 2 && (quote === "\"" || quote === "'") && value.endsWith(quote)) {
    return value.slice(1, -1);
  }
  return value;
}
